File: guajira-bag/API/GaleriaService.cs
using GuajiraBags.Api.Data;
using GuajiraBags.Api.Models;
using GuajiraBags.Api.Models.DTOs;
using Microsoft.EntityFrameworkCore;

namespace GuajiraBags.Api.Services;

public class GaleriaService(AppDbContext db, FileStorageService storage)
{
    // ── Listar ───────────────────────────────────────────────
    public async Task<List<GaleriaResponse>> ListarAsync(bool soloActivos = true)
    {
        var query = db.Galeria.AsQueryable();
        if (soloActivos) query = query.Where(g => g.Activo);

        return await query
            .OrderBy(g => g.Orden)
            .ThenByDescending(g => g.CreadoEn)
            .Select(g => ToResponse(g))
            .ToListAsync();
    }

    // ── Obtener por Id ───────────────────────────────────────
    public async Task<GaleriaResponse?> ObtenerAsync(int id)
    {
        var g = await db.Galeria.FindAsync(id);
        return g is null ? null : ToResponse(g);
    }

    // ── Crear (con foto obligatoria) ─────────────────────────
    public async Task<(bool Ok, string Error, GaleriaResponse? Data)> CrearAsync(
        IFormFile foto, GaleriaCreateRequest req)
    {
        var (ok, error, url) = await storage.SaveAsync(foto, "galeria");
        if (!ok) return (false, error, null);

        var maxOrden = await db.Galeria.MaxAsync(g => (short?)g.Orden) ?? 0;

        var item = new GaleriaItem
        {
            Url     = url,
            Caption = req.Caption,
            Orden   = (short)(maxOrden + 1)
        };

        db.Galeria.Add(item);
        await db.SaveChangesAsync();
        return (true, string.Empty, ToResponse(item));
    }

    // ── Actualizar caption / orden / activo ──────────────────
    public async Task<(bool Ok, string Error, GaleriaResponse? Data)> ActualizarAsync(
        int id, GaleriaUpdateRequest req)
    {
        var item = await db.Galeria.FindAsync(id);
        if (item is null) return (false, "Imagen no encontrada", null);

        if (req.Caption is not null) item.Caption = req.Caption;
        if (req.Orden.HasValue)      item.Orden   = req.Orden.Value;
        if (req.Activo.HasValue)     item.Activo  = req.Activo.Value;

        await db.SaveChangesAsync();
        return (true, string.Empty, ToResponse(item));
    }

    // ── Eliminar ─────────────────────────────────────────────
    public async Task<bool> EliminarAsync(int id)
    {
        var item = await db.Galeria.FindAsync(id);
        if (item is null) return false;

        storage.Delete(item.Url);
        db.Galeria.Remove(item);
        await db.SaveChangesAsync();
        return true;
    }

    // ── Mapper ───────────────────────────────────────────────
    private static GaleriaResponse ToResponse(GaleriaItem g) =>
        new(g.Id, g.Url, g.Caption, g.Orden, g.Activo, g.CreadoEn);
}

File: Nueva carpeta/php-api/src/Services/GaleriaService.php
<?php

namespace GuajiraBags\Services;

use GuajiraBags\Helpers\Response;
use PDO;

class GaleriaService
{
    public function __construct(
        private PDO $db,
        private FileStorageService $storage
    ) {}

    public function listar(bool $soloActivos = true): array
    {
        $sql = 'SELECT * FROM galeria';
        if ($soloActivos) {
            $sql .= ' WHERE activo = 1';
        }
        $sql .= ' ORDER BY orden ASC, creado_en DESC';

        $rows = $this->db->query($sql)->fetchAll();
        return array_map(fn($row) => $this->toResponse($row), $rows);
    }

    public function obtener(int $id): ?array
    {
        $stmt = $this->db->prepare('SELECT * FROM galeria WHERE id = ?');
        $stmt->execute([$id]);
        $row = $stmt->fetch();
        return $row ? $this->toResponse($row) : null;
    }

    /** @return array{0: bool, 1: string, 2: ?array} */
    public function crear(array $foto, ?string $caption): array
    {
        [$ok, $error, $url] = $this->storage->save($foto, 'galeria');
        if (!$ok) {
            return [false, $error, null];
        }

        $maxOrden = (int) $this->db->query('SELECT COALESCE(MAX(orden), 0) FROM galeria')->fetchColumn();

        $stmt = $this->db->prepare(
            'INSERT INTO galeria (url, caption, orden) VALUES (?, ?, ?)'
        );
        $stmt->execute([$url, $caption, $maxOrden + 1]);

        return [true, '', $this->obtener((int) $this->db->lastInsertId())];
    }

    /** @return array{0: bool, 1: string, 2: ?array} */
    public function actualizar(int $id, array $req): array
    {
        $stmt = $this->db->prepare('SELECT * FROM galeria WHERE id = ?');
        $stmt->execute([$id]);
        $item = $stmt->fetch();

        if (!$item) {
            return [false, 'Imagen no encontrada', null];
        }

        $fields = [];
        $params = [];

        if (array_key_exists('caption', $req)) {
            $fields[] = 'caption = ?';
            $params[] = $req['caption'];
        }
        if (isset($req['orden'])) {
            $fields[] = 'orden = ?';
            $params[] = (int) $req['orden'];
        }
        if (isset($req['activo'])) {
            $fields[] = 'activo = ?';
            $params[] = $req['activo'] ? 1 : 0;
        }

        if (!empty($fields)) {
            $params[] = $id;
            $sql = 'UPDATE galeria SET ' . implode(', ', $fields) . ' WHERE id = ?';
            $stmt = $this->db->prepare($sql);
            $stmt->execute($params);
        }

        return [true, '', $this->obtener($id)];
    }

    public function eliminar(int $id): bool
    {
        $stmt = $this->db->prepare('SELECT * FROM galeria WHERE id = ?');
        $stmt->execute([$id]);
        $item = $stmt->fetch();

        if (!$item) {
            return false;
        }

        $this->storage->delete($item['url']);
        $stmt = $this->db->prepare('DELETE FROM galeria WHERE id = ?');
        $stmt->execute([$id]);
        return true;
    }

    private function toResponse(array $row): array
    {
        return [
            'id'       => (int) $row['id'],
            'url'      => $row['url'],
            'caption'  => $row['caption'],
            'orden'    => (int) $row['orden'],
            'activo'   => Response::boolFromDb($row['activo']),
            'creadoEn' => $row['creado_en']
                ? (new \DateTimeImmutable($row['creado_en']))->format('Y-m-d\TH:i:s')
                : null,
        ];
    }
}
